### PR TITLE
Improve rpad udf by using a GenericStringBuilder

### LINKED ISSUE #11997: Speed up rpad implementation to use StringBuilder
# Usecase

@Omega359  made a nice improvement to `lpad` in https://github.com/apache/datafusion/pull/11987 to make the code cleaner and faster. 

This ticket tracks the work to apply the same treatment of https://github.com/apache/datafusion/pull/11987 to `rpad` -- which @Lordworms ported in https://github.com/apache/datafusion/pull/11942

              rpad update with utf8view wasn't committed yet. I know it would work anyways but I think now that rpad update is committed it's good to uncomment.

_Originally posted by @Omega359 in https://github.com/apache/datafusion/pull/11987#discussion_r1717559901_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #11997

### CODE REVIEW

**Comment by Lordworms** [comment]
Sorry for the late update, got some issue last week

**Review by alamb — APPROVED** [review]
Likewise this looks good to me -- thank you @Lordworms 

We just need to sort out the `cargo fmt` CI Error

maybe we could make two separate macros? Or use `StringArrayType` to make generic functions instead -- see https://github.com/apache/datafusion/pull/12027

**Comment by Lordworms** [comment]
> 
Sure

File: datafusion/functions/src/unicode/rpad.rs [hunk 1]
```diff
@@ -84,170 +87,182 @@ impl ScalarUDFImpl for RPadFunc {
...
-            .map(|((string, length), fill)| match (string, length, fill) {
-                (Some(string), Some(length), Some(fill)) => {
-                    if length > i32::MAX as i64 {
-                        return exec_err!("rpad requested length {} too large", length);
-                    }
-
-                    let length = if length < 0 { 0 } else { length as usize };
-                    let graphemes = string.graphemes(true).collect::<Vec<&str>>();
-                    let fill_chars = fill.chars().collect::<Vec<char>>();
+pub fn rpad<StringArrayLen: OffsetSizeTrait, FillArrayLen: OffsetSizeTrait>(
```
Comment (alamb) [inline datafusion/functions/src/unicode/rpad.rs:118]: 😍  -- 100% for no more macros!

**Review by alamb — APPROVED** [review]
Awesome -- very nice @Lordworms 

Thank you for the review and help @Omega359

**Pull request merged**